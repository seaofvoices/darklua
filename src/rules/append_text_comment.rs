use std::fs;
use std::path::{Path, PathBuf};
use std::sync::OnceLock;

use crate::nodes::{
    Block, BlockTokens, DoTokens, FunctionBodyTokens, GenericForTokens, Identifier,
    IfStatementTokens, LastStatement, LocalAssignTokens, LocalFunctionTokens, NumericForTokens,
    ParentheseExpression, ParentheseTokens, Prefix, RepeatTokens, ReturnTokens, Statement, Token,
    TriviaKind, TypeDeclarationTokens, Variable, WhileTokens,
};
use crate::rules::{
    verify_property_collisions, verify_required_any_properties, Context, Rule, RuleConfiguration,
    RuleConfigurationError, RuleProcessResult, RuleProperties,
};

use super::{FlawlessRule, ShiftTokenLine};

pub const APPEND_TEXT_COMMENT_RULE_NAME: &str = "append_text_comment";

/// A rule to append a comment at the beginning or the end of each file.
#[derive(Debug, Default)]
pub struct AppendTextComment {
    text_value: OnceLock<Result<String, String>>,
    text_content: TextContent,
    location: AppendLocation,
}

impl AppendTextComment {
    pub fn new(value: impl Into<String>) -> Self {
        Self {
            text_value: Default::default(),
            text_content: TextContent::Value(value.into()),
            location: Default::default(),
        }
    }

    pub fn from_file_content(file_path: impl Into<PathBuf>) -> Self {
        Self {
            text_value: Default::default(),
            text_content: TextContent::FilePath(file_path.into()),
            location: Default::default(),
        }
    }

    pub fn at_end(mut self) -> Self {
        self.location = AppendLocation::End;
        self
    }

    fn text(&self, project_path: &Path) -> Result<String, String> {
        self.text_value
            .get_or_init(|| {
                match &self.text_content {
                    TextContent::None => Err("".to_owned()),
                    TextContent::Value(value) => Ok(value.clone()),
                    TextContent::FilePath(file_path) => {
                        fs::read_to_string(project_path.join(file_path)).map_err(|err| {
                            format!("unable to read file `{}`: {}", file_path.display(), err)
                        })
                    }
                }
                .map(|content| {
                    if content.is_empty() {
                        "".to_owned()
                    } else if content.contains('\n') {
                        let mut equal_count = 0;

                        let close_comment = loop {
                            let close_comment = format!("]{}]", "=".repeat(equal_count));
                            if !content.contains(&close_comment) {
                                break close_comment;
                            }
                            equal_count += 1;
                        };

                        format!(
                            "--[{}[\n{}\n{}",
                            "=".repeat(equal_count),
                            content,
                            close_comment
                        )
                    } else {
                        format!("--{}", content)
                    }
                })
            })
            .clone()
    }
}

impl Rule for AppendTextComment {
    fn process(&self, block: &mut Block, context: &Context) -> RuleProcessResult {
        let text = self.text(context.project_location())?;

        if text.is_empty() {
            return Ok(());
        }

        let shift_lines = text.lines().count();
        ShiftTokenLine::new(shift_lines as isize).flawless_process(block, context);

        match self.location {
            AppendLocation::Start => {
                if let Some(statement) = block.first_mut_statement() {
                    match statement {
                        Statement::Assign(assign_statement) => {
                            let variable = assign_statement
                                .iter_mut_variables()
                                .next()
                                .ok_or("an assign statement must have at least one variable")?;
                            self.location
                                .append_comment(variable_get_first_token(variable), text);
                        }
                        Statement::Do(do_statement) => {
                            if let Some(tokens) = do_statement.mutate_tokens() {
                                self.location.append_comment(&mut tokens.r#do, text);
                            } else {
                                let mut token = Token::from_content("do");
                                self.location.append_comment(&mut token, text);

                                do_statement.set_tokens(DoTokens {
                                    r#do: token,
                                    end: Token::from_content("end"),
                                });
                            }
                        }
                        Statement::Call(call) => {
                            self.location
                                .append_comment(prefix_get_first_token(call.mutate_prefix()), text);
                        }
                        Statement::CompoundAssign(compound_assign) => {
                            self.location.append_comment(
                                variable_get_first_token(compound_assign.mutate_variable()),
                                text,
                            );
                        }
                        Statement::Function(function) => {
                            if let Some(tokens) = function.mutate_tokens() {
                                self.location.append_comment(&mut tokens.function, text);
                            } else {
                                let mut token = Token::from_content("function");
                                self.location.append_comment(&mut token, text);

                                function.set_tokens(FunctionBodyTokens {
                                    function: token,
                                    opening_parenthese: Token::from_content("("),
                                    closing_parenthese: Token::from_content(")"),
                                    end: Token::from_content("end"),
                                    parameter_commas: Vec::new(),
                                    variable_arguments: None,
                                    variable_arguments_colon: None,
                                    return_type_colon: None,
                                });
                            }
                        }
                        Statement::GenericFor(generic_for) => {
                            if let Some(tokens) = generic_for.mutate_tokens() {
                                self.location.append_comment(&mut tokens.r#for, text);
                            } else {
                                let mut token = Token::from_content("for");
                                self.location.append_comment(&mut token, text);

                                generic_for.set_tokens(GenericForTokens {
                                    r#for: token,
                                    r#in: Token::from_content("in"),
                                    r#do: Token::from_content("do"),
                                    end: Token::from_content("end"),
                                    identifier_commas: Vec::new(),
                                    value_commas: Vec::new(),
                                });
                            }
                        }
                        Statement::If(if_statement) => {
                            if let Some(tokens) = if_statement.mutate_tokens() {
                                self.location.append_comment(&mut tokens.r#if, text);
                            } else {
                                let mut token = Token::from_content("if");
                                self.location.append_comment(&mut token, text);

                                if_statement.set_tokens(IfStatementTokens {
                                    r#if: token,
                                    then: Token::from_content("then"),
                                    end: Token::from_content("end"),
                                    r#else: None,
                                });
                            }
                        }
                        Statement::LocalAssign(local_assign) => {
                            if let Some(tokens) = local_assign.mutate_tokens() {
                                self.location.append_comment(&mut tokens.local, text);
                            } else {
                                let mut token = Token::from_content("local");
                                self.location.append_comment(&mut token, text);

                                local_assign.set_tokens(LocalAssignTokens {
                                    local: token,
                                    equal: None,
                                    variable_commas: Vec::new(),
                                    value_commas: Vec::new(),
                                });
                            }
                        }
                        Statement::LocalFunction(local_function) => {
                            if let Some(tokens) = local_function.mutate_tokens() {
                                self.location.append_comment(&mut tokens.local, text);
                            } else {
                                let mut token = Token::from_content("local");
                                self.location.append_comment(&mut token, text);

                                local_function.set_tokens(LocalFunctionTokens {
                                    local: token,
                                    function_body: FunctionBodyTokens {
                                        function: Token::from_content("function"),
                                        opening_parenthese: Token::from_content("("),
                                        closing_parenthese: Token::from_content(")"),
                                        end: Token::from_content("end"),
                                        parameter_commas: Vec::new(),
                                        variable_arguments: None,
                                        variable_arguments_colon: None,
                                        return_type_colon: None,
                                    },
                                });
                            }
                        }
                        Statement::NumericFor(numeric_for) => {
                            if let Some(tokens) = numeric_for.mutate_tokens() {
                                self.location.append_comment(&mut tokens.r#for, text);
                            } else {
                                let mut token = Token::from_content("for");
                                self.location.append_comment(&mut token, text);

                                numeric_for.set_tokens(NumericForTokens {
                                    r#for: token,
                                    equal: Token::from_content("="),
                                    r#do: Token::from_content("do"),
                                    end: Token::from_content("end"),
                                    end_comma: Token::from_content(","),
                                    step_comma: None,
                                });
                            }
                        }
                        Statement::Repeat(repeat) => {
                            if let Some(tokens) = repeat.mutate_tokens() {
                                self.location.append_comment(&mut tokens.repeat, text);
                            } else {
                                let mut token = Token::from_content("repeat");
                                self.location.append_comment(&mut token, text);

                                repeat.set_tokens(RepeatTokens {
                                    repeat: token,
                                    until: Token::from_content("until"),
                                });
                            }
                        }
                        Statement::While(while_statement) => {
                            if let Some(tokens) = while_statement.mutate_tokens() {
                                self.location.append_comment(&mut tokens.r#while, text);
                            } else {
                                let mut token = Token::from_content("while");
                                self.location.append_comment(&mut token, text);

                                while_statement.set_tokens(WhileTokens {
                                    r#while: token,
                                    r#do: Token::from_content("do"),
                                    end: Token::from_content("end"),
                                });
                            }
                        }
                        Statement::TypeDeclaration(type_declaration) => {
                            let is_exported = type_declaration.is_exported();
                            if let Some(tokens) = type_declaration.mutate_tokens() {
                                if is_exported {
                                    self.location.append_comment(
                                        tokens
                                            .export
                                            .get_or_insert_with(|| Token::from_content("export")),
                                        text,
                                    );
                                } else {
                                    self.location.append_comment(&mut tokens.r#type, text);
                                }
                            } else if is_exported {
                                let mut token = Token::from_content("export");
                                self.location.append_comment(&mut token, text);

                                type_declaration.set_tokens(TypeDeclarationTokens {
                                    r#type: Token::from_content("type"),
                                    equal: Token::from_content("="),
                                    export: Some(token),
                                });
                            } else {
                                let mut token = Token::from_content("type");
                                self.location.append_comment(&mut token, text);

                                type_declaration.set_tokens(TypeDeclarationTokens {
                                    r#type: token,
                                    equal: Token::from_content("="),
                                    export: None,
                                });
                            }
                        }
                    }
                } else if let Some(statement) = block.mutate_last_statement() {
                    match statement {
                        LastStatement::Break(token) => {
                            self.location.append_comment(
                                token.get_or_insert_with(|| Token::from_content("break")),
                                text,
                            );
                        }
                        LastStatement::Continue(token) => {
                            self.location.append_comment(
                                token.get_or_insert_with(|| Token::from_content("continue")),
                                text,
                            );
                        }
                        LastStatement::Return(return_statement) => {
                            if let Some(tokens) = return_statement.mutate_tokens() {
                                self.location.append_comment(&mut tokens.r#return, text);
                            } else {
                                let mut token = Token::from_content("return");
                                self.location.append_comment(&mut token, text);

                                return_statement.set_tokens(ReturnTokens {
                                    r#return: token,
                                    commas: Vec::new(),
                                });
                            }
                        }
                    }
                } else {
                    self.location.write_to_block(block, text);
                }
            }
            AppendLocation::End => {
                self.location.write_to_block(block, text);
            }
        }

        Ok(())
    }
}

fn variable_get_first_token(variable: &mut Variable) -> &mut Token {
    match variable {
        Variable::Identifier(identifier) => identifier_get_first_token(identifier),
        Variable::Field(field_expression) => {
            prefix_get_first_token(field_expression.mutate_prefix())
        }
        Variable::Index(index_expression) => {
            prefix_get_first_token(index_expression.mutate_prefix())
        }
    }
}

fn prefix_get_first_token(prefix: &mut Prefix) -> &mut Token {
    let mut current = prefix;
    loop {
        match current {
            Prefix::Call(call) => {
                current = call.mutate_prefix();
            }
            Prefix::Field(field_expression) => {
                current = field_expression.mutate_prefix();
            }
            Prefix::Index(index_expression) => {
                current = index_expression.mutate_prefix();
            }
            Prefix::Identifier(identifier) => break identifier_get_first_token(identifier),
            Prefix::Parenthese(parenthese_expression) => {
                break parentheses_get_first_token(parenthese_expression)
            }
        }
    }
}

fn identifier_get_first_token(identifier: &mut Identifier) -> &mut Token {
    if identifier.get_token().is_none() {
        let name = identifier.get_name().to_owned();
        identifier.set_token(Token::from_content(name));
    }
    identifier.mutate_token().unwrap()
}

fn parentheses_get_first_token(parentheses: &mut ParentheseExpression) -> &mut Token {
    if parentheses.get_tokens().is_none() {
        parentheses.set_tokens(ParentheseTokens {
            left_parenthese: Token::from_content("("),
            right_parenthese: Token::from_content(")"),
        });
    }
    &mut parentheses.mutate_tokens().unwrap().left_parenthese
}

impl RuleConfiguration for AppendTextComment {
    fn configure(&mut self, properties: RuleProperties) -> Result<(), RuleConfigurationError> {
        verify_required_any_properties(&properties, &["text", "file"])?;
        verify_property_collisions(&properties, &["text", "file"])?;

        for (key, value) in properties {
            match key.as_str() {
                "text" => {
                    self.text_content = TextContent::Value(value.expect_string(&key)?);
                }
                "file" => {
                    self.text_content =
                        TextContent::FilePath(PathBuf::from(value.expect_string(&key)?));
                }
                "location" => {
                    self.location = match value.expect_string(&key)?.as_str() {
                        "start" => AppendLocation::Start,
                        "end" => AppendLocation::End,
                        unexpected => {
                            return Err(RuleConfigurationError::UnexpectedValue {
                                property: "location".to_owned(),
                                message: format!(
                                    "invalid value `{}` (must be `start` or `end`)",
                                    unexpected
                                ),
                            })
                        }
                    };
                }
                _ => return Err(RuleConfigurationError::UnexpectedProperty(key)),
            }
        }

        Ok(())
    }

    fn get_name(&self) -> &'static str {
        APPEND_TEXT_COMMENT_RULE_NAME
    }

    fn serialize_to_properties(&self) -> RuleProperties {
        let mut properties = RuleProperties::new();

        match self.location {
            AppendLocation::Start => {}
            AppendLocation::End => {
                properties.insert("location".to_owned(), "end".into());
            }
        }

        match &self.text_content {
            TextContent::None => {}
            TextContent::Value(value) => {
                properties.insert("text".to_owned(), value.into());
            }
            TextContent::FilePath(file_path) => {
                properties.insert(
                    "file".to_owned(),
                    file_path.to_string_lossy().to_string().into(),
                );
            }
        }

        properties
    }
}

#[derive(Debug, PartialEq, Eq)]
enum TextContent {
    None,
    Value(String),
    FilePath(PathBuf),
}

impl Default for TextContent {
    fn default() -> Self {
        Self::None
    }
}

#[derive(Debug, PartialEq, Eq)]
enum AppendLocation {
    Start,
    End,
}

impl AppendLocation {
    fn write_to_block(&self, block: &mut Block, comment: String) {
        if let Some(tokens) = block.mutate_tokens() {
            let final_token = tokens
                .final_token
                .get_or_insert_with(|| Token::from_content(""));
            self.append_comment(final_token, comment);
        } else {
            let mut token = Token::from_content("");
            self.append_comment(&mut token, comment);

            block.set_tokens(BlockTokens {
                semicolons: Vec::new(),
                last_semicolon: None,
                final_token: Some(token),
            });
        }
    }

    fn append_comment(&self, token: &mut Token, comment: String) {
        match self {
            AppendLocation::Start => {
                token.push_leading_trivia(TriviaKind::Comment.with_content(comment));
            }
            AppendLocation::End => {
                token.push_trailing_trivia(TriviaKind::Comment.with_content(comment));
            }
        }
    }
}

impl Default for AppendLocation {
    fn default() -> Self {
        Self::Start
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use crate::rules::Rule;

    use insta::assert_json_snapshot;

    #[test]
    fn serialize_rule_with_text() {
        let rule: Box<dyn Rule> = Box::new(AppendTextComment::new("content"));

        assert_json_snapshot!("append_text_comment_with_text", rule);
    }

    #[test]
    fn serialize_rule_with_text_at_end() {
        let rule: Box<dyn Rule> = Box::new(AppendTextComment::new("content").at_end());

        assert_json_snapshot!("append_text_comment_with_text_at_end", rule);
    }

    #[test]
    fn configure_with_extra_field_error() {
        let result = json5::from_str::<Box<dyn Rule>>(
            r#"{
            rule: 'append_text_comment',
            text: '',
            prop: "something",
        }"#,
        );
        pretty_assertions::assert_eq!(result.unwrap_err().to_string(), "unexpected field 'prop'");
    }
}
