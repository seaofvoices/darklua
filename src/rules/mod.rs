//! A module that contains the different rules that mutates a Lua block.

mod append_text_comment;
pub mod bundle;
mod call_parens;
mod compute_expression;
mod configuration_error;
mod convert_index_to_field;
mod convert_require;
mod empty_do;
mod filter_early_return;
mod group_local;
mod inject_value;
mod method_def;
mod no_local_function;
mod remove_assertions;
mod remove_call_match;
mod remove_comments;
mod remove_compound_assign;
mod remove_debug_profiling;
mod remove_interpolated_string;
mod remove_nil_declarations;
mod remove_spaces;
mod remove_types;
mod remove_unused_variable;
mod rename_variables;
mod replace_referenced_tokens;
pub(crate) mod require;
mod rule_property;
mod shift_token_line;
mod unused_if_branch;
mod unused_while;

pub use append_text_comment::*;
pub use call_parens::*;
pub use compute_expression::*;
pub use configuration_error::RuleConfigurationError;
pub use convert_index_to_field::*;
pub use convert_require::*;
pub use empty_do::*;
pub use filter_early_return::*;
pub use group_local::*;
pub use inject_value::*;
pub use method_def::*;
pub use no_local_function::*;
pub use remove_assertions::*;
pub use remove_comments::*;
pub use remove_compound_assign::*;
pub use remove_debug_profiling::*;
pub use remove_interpolated_string::*;
pub use remove_nil_declarations::*;
pub use remove_spaces::*;
pub use remove_types::*;
pub use remove_unused_variable::*;
pub use rename_variables::*;
pub(crate) use replace_referenced_tokens::*;
pub use rule_property::*;
pub(crate) use shift_token_line::*;
pub use unused_if_branch::*;
pub use unused_while::*;

use crate::nodes::Block;
use crate::Resources;

use serde::de::{self, MapAccess, Visitor};
use serde::ser::SerializeMap;
use serde::{Deserialize, Deserializer, Serialize, Serializer};
use std::collections::HashMap;
use std::fmt;
use std::path::{Path, PathBuf};
use std::str::FromStr;

#[derive(Debug, Clone)]
pub struct ContextBuilder<'a, 'resources, 'code> {
    path: PathBuf,
    resources: &'resources Resources,
    original_code: &'code str,
    blocks: HashMap<PathBuf, &'a Block>,
    project_location: Option<PathBuf>,
}

impl<'a, 'resources, 'code> ContextBuilder<'a, 'resources, 'code> {
    pub fn new(
        path: impl Into<PathBuf>,
        resources: &'resources Resources,
        original_code: &'code str,
    ) -> Self {
        Self {
            path: path.into(),
            resources,
            original_code,
            blocks: Default::default(),
            project_location: None,
        }
    }

    pub fn with_project_location(mut self, path: impl Into<PathBuf>) -> Self {
        self.project_location = Some(path.into());
        self
    }

    pub fn build(self) -> Context<'a, 'resources, 'code> {
        Context {
            path: self.path,
            resources: self.resources,
            original_code: self.original_code,
            blocks: self.blocks,
            project_location: self.project_location,
        }
    }

    pub fn insert_block<'block: 'a>(&mut self, path: impl Into<PathBuf>, block: &'block Block) {
        self.blocks.insert(path.into(), block);
    }
}

/// The intent of this struct is to hold data shared across all rules applied to a file.
#[derive(Debug, Clone)]
pub struct Context<'a, 'resources, 'code> {
    path: PathBuf,
    resources: &'resources Resources,
    original_code: &'code str,
    blocks: HashMap<PathBuf, &'a Block>,
    project_location: Option<PathBuf>,
}

impl<'a, 'resources, 'code> Context<'a, 'resources, 'code> {
    pub fn block(&self, path: impl AsRef<Path>) -> Option<&Block> {
        self.blocks.get(path.as_ref()).copied()
    }

    pub fn current_path(&self) -> &Path {
        self.path.as_ref()
    }

    fn resources(&self) -> &Resources {
        self.resources
    }

    fn original_code(&self) -> &str {
        self.original_code
    }

    fn project_location(&self) -> &Path {
        self.project_location.as_deref().unwrap_or_else(|| {
            let source = self.current_path();
            source.parent().unwrap_or_else(|| {
                log::warn!(
                    "unexpected file path `{}` (unable to extract parent path)",
                    source.display()
                );
                source
            })
        })
    }
}

pub type RuleProcessResult = Result<(), String>;

/// Defines an interface that will be used to mutate blocks and how to serialize and deserialize
/// the rule configuration.
pub trait Rule: RuleConfiguration + fmt::Debug {
    /// This method should mutate the given block to apply the rule
    fn process(&self, block: &mut Block, context: &Context) -> RuleProcessResult;

    /// Return the list of paths to Lua files that is necessary to apply this rule. This will load
    /// each AST block from these files into the context object.
    fn require_content(&self, _current_source: &Path, _current_block: &Block) -> Vec<PathBuf> {
        Vec::new()
    }
}

pub trait RuleConfiguration {
    /// The rule deserializer will construct the default rule and then send the properties through
    /// this method to modify the behavior of the rule.
    fn configure(&mut self, properties: RuleProperties) -> Result<(), RuleConfigurationError>;
    /// This method should return the unique name of the rule.
    fn get_name(&self) -> &'static str;
    /// For implementing the serialize trait on the Rule trait, this method should return all
    /// properties that differs from their default value.
    fn serialize_to_properties(&self) -> RuleProperties;
    /// Returns `true` if the rule has at least one property.
    fn has_properties(&self) -> bool {
        !self.serialize_to_properties().is_empty()
    }
}

pub trait FlawlessRule {
    fn flawless_process(&self, block: &mut Block, context: &Context);
}

impl<T: FlawlessRule + RuleConfiguration + fmt::Debug> Rule for T {
    fn process(&self, block: &mut Block, context: &Context) -> RuleProcessResult {
        self.flawless_process(block, context);
        Ok(())
    }
}

/// A function to get the default rule stack for darklua. All the rules here must preserve all the
/// functionalities of the original code after being applied. They must guarantee that the
/// processed block will work as much as the original one.
pub fn get_default_rules() -> Vec<Box<dyn Rule>> {
    vec![
        Box::<RemoveSpaces>::default(),
        Box::<RemoveComments>::default(),
        Box::<ComputeExpression>::default(),
        Box::<RemoveUnusedIfBranch>::default(),
        Box::<RemoveUnusedWhile>::default(),
        Box::<FilterAfterEarlyReturn>::default(),
        Box::<RemoveEmptyDo>::default(),
        Box::<RemoveUnusedVariable>::default(),
        Box::<RemoveMethodDefinition>::default(),
        Box::<ConvertIndexToField>::default(),
        Box::<RemoveNilDeclaration>::default(),
        Box::<RenameVariables>::default(),
        Box::<RemoveFunctionCallParens>::default(),
    ]
}

pub fn get_all_rule_names() -> Vec<&'static str> {
    vec![
        APPEND_TEXT_COMMENT_RULE_NAME,
        COMPUTE_EXPRESSIONS_RULE_NAME,
        CONVERT_INDEX_TO_FIELD_RULE_NAME,
        CONVERT_LOCAL_FUNCTION_TO_ASSIGN_RULE_NAME,
        CONVERT_REQUIRE_RULE_NAME,
        FILTER_AFTER_EARLY_RETURN_RULE_NAME,
        GROUP_LOCAL_ASSIGNMENT_RULE_NAME,
        INJECT_GLOBAL_VALUE_RULE_NAME,
        REMOVE_ASSERTIONS_RULE_NAME,
        REMOVE_COMMENTS_RULE_NAME,
        REMOVE_COMPOUND_ASSIGNMENT_RULE_NAME,
        REMOVE_DEBUG_PROFILING_RULE_NAME,
        REMOVE_EMPTY_DO_RULE_NAME,
        REMOVE_FUNCTION_CALL_PARENS_RULE_NAME,
        REMOVE_INTERPOLATED_STRING_RULE_NAME,
        REMOVE_METHOD_DEFINITION_RULE_NAME,
        REMOVE_NIL_DECLARATION_RULE_NAME,
        REMOVE_SPACES_RULE_NAME,
        REMOVE_TYPES_RULE_NAME,
        REMOVE_UNUSED_IF_BRANCH_RULE_NAME,
        REMOVE_UNUSED_VARIABLE_RULE_NAME,
        REMOVE_UNUSED_WHILE_RULE_NAME,
        RENAME_VARIABLES_RULE_NAME,
    ]
}

impl FromStr for Box<dyn Rule> {
    type Err = String;

    fn from_str(string: &str) -> Result<Self, Self::Err> {
        let rule: Box<dyn Rule> = match string {
            APPEND_TEXT_COMMENT_RULE_NAME => Box::<AppendTextComment>::default(),
            COMPUTE_EXPRESSIONS_RULE_NAME => Box::<ComputeExpression>::default(),
            CONVERT_INDEX_TO_FIELD_RULE_NAME => Box::<ConvertIndexToField>::default(),
            CONVERT_LOCAL_FUNCTION_TO_ASSIGN_RULE_NAME => {
                Box::<ConvertLocalFunctionToAssign>::default()
            }
            CONVERT_REQUIRE_RULE_NAME => Box::<ConvertRequire>::default(),
            FILTER_AFTER_EARLY_RETURN_RULE_NAME => Box::<FilterAfterEarlyReturn>::default(),
            GROUP_LOCAL_ASSIGNMENT_RULE_NAME => Box::<GroupLocalAssignment>::default(),
            INJECT_GLOBAL_VALUE_RULE_NAME => Box::<InjectGlobalValue>::default(),
            REMOVE_ASSERTIONS_RULE_NAME => Box::<RemoveAssertions>::default(),
            REMOVE_COMMENTS_RULE_NAME => Box::<RemoveComments>::default(),
            REMOVE_COMPOUND_ASSIGNMENT_RULE_NAME => Box::<RemoveCompoundAssignment>::default(),
            REMOVE_DEBUG_PROFILING_RULE_NAME => Box::<RemoveDebugProfiling>::default(),
            REMOVE_EMPTY_DO_RULE_NAME => Box::<RemoveEmptyDo>::default(),
            REMOVE_FUNCTION_CALL_PARENS_RULE_NAME => Box::<RemoveFunctionCallParens>::default(),
            REMOVE_INTERPOLATED_STRING_RULE_NAME => Box::<RemoveInterpolatedString>::default(),
            REMOVE_METHOD_DEFINITION_RULE_NAME => Box::<RemoveMethodDefinition>::default(),
            REMOVE_NIL_DECLARATION_RULE_NAME => Box::<RemoveNilDeclaration>::default(),
            REMOVE_SPACES_RULE_NAME => Box::<RemoveSpaces>::default(),
            REMOVE_TYPES_RULE_NAME => Box::<RemoveTypes>::default(),
            REMOVE_UNUSED_IF_BRANCH_RULE_NAME => Box::<RemoveUnusedIfBranch>::default(),
            REMOVE_UNUSED_VARIABLE_RULE_NAME => Box::<RemoveUnusedVariable>::default(),
            REMOVE_UNUSED_WHILE_RULE_NAME => Box::<RemoveUnusedWhile>::default(),
            RENAME_VARIABLES_RULE_NAME => Box::<RenameVariables>::default(),
            _ => return Err(format!("invalid rule name: {}", string)),
        };

        Ok(rule)
    }
}

impl Serialize for dyn Rule {
    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
        let properties = self.serialize_to_properties();
        let property_count = properties.len();
        let rule_name = self.get_name();

        if property_count == 0 {
            serializer.serialize_str(rule_name)
        } else {
            let mut map = serializer.serialize_map(Some(property_count + 1))?;

            map.serialize_entry("rule", rule_name)?;

            let mut ordered: Vec<(String, RulePropertyValue)> = properties.into_iter().collect();

            ordered.sort_by(|a, b| a.0.cmp(&b.0));

            for (key, value) in ordered {
                map.serialize_entry(&key, &value)?;
            }

            map.end()
        }
    }
}

impl<'de> Deserialize<'de> for Box<dyn Rule> {
    fn deserialize<D: Deserializer<'de>>(deserializer: D) -> Result<Box<dyn Rule>, D::Error> {
        struct StringOrStruct;

        impl<'de> Visitor<'de> for StringOrStruct {
            type Value = Box<dyn Rule>;

            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
                formatter.write_str("rule name or rule object")
            }

            fn visit_str<E>(self, value: &str) -> Result<Self::Value, E>
            where
                E: de::Error,
            {
                let mut rule: Self::Value = FromStr::from_str(value).map_err(de::Error::custom)?;

                rule.configure(RuleProperties::new())
                    .map_err(de::Error::custom)?;

                Ok(rule)
            }

            fn visit_map<M>(self, mut map: M) -> Result<Self::Value, M::Error>
            where
                M: MapAccess<'de>,
            {
                let mut rule_name = None;
                let mut properties = HashMap::new();

                while let Some(key) = map.next_key::<String>()? {
                    match key.as_str() {
                        "rule" => {
                            if rule_name.is_none() {
                                rule_name.replace(map.next_value::<String>()?);
                            } else {
                                return Err(de::Error::duplicate_field("rule"));
                            }
                        }
                        property => {
                            let value = map.next_value::<RulePropertyValue>()?;

                            if properties.insert(property.to_owned(), value).is_some() {
                                return Err(de::Error::custom(format!(
                                    "duplicate field {} in rule object",
                                    property
                                )));
                            }
                        }
                    }
                }

                if let Some(rule_name) = rule_name {
                    let mut rule: Self::Value =
                        FromStr::from_str(&rule_name).map_err(de::Error::custom)?;

                    rule.configure(properties).map_err(de::Error::custom)?;

                    Ok(rule)
                } else {
                    Err(de::Error::missing_field("rule"))
                }
            }
        }

        deserializer.deserialize_any(StringOrStruct)
    }
}

fn verify_no_rule_properties(properties: &RuleProperties) -> Result<(), RuleConfigurationError> {
    if let Some((key, _value)) = properties.iter().next() {
        return Err(RuleConfigurationError::UnexpectedProperty(key.to_owned()));
    }
    Ok(())
}

fn verify_required_properties(
    properties: &RuleProperties,
    names: &[&str],
) -> Result<(), RuleConfigurationError> {
    for name in names.iter() {
        if !properties.contains_key(*name) {
            return Err(RuleConfigurationError::MissingProperty(name.to_string()));
        }
    }
    Ok(())
}

fn verify_required_any_properties(
    properties: &RuleProperties,
    names: &[&str],
) -> Result<(), RuleConfigurationError> {
    if names.iter().any(|name| properties.contains_key(*name)) {
        Ok(())
    } else {
        Err(RuleConfigurationError::MissingAnyProperty(
            names.iter().map(ToString::to_string).collect(),
        ))
    }
}

fn verify_property_collisions(
    properties: &RuleProperties,
    names: &[&str],
) -> Result<(), RuleConfigurationError> {
    let mut exists = false;
    for name in names.iter() {
        if properties.contains_key(*name) {
            if exists {
                return Err(RuleConfigurationError::PropertyCollision(
                    names.iter().map(ToString::to_string).collect(),
                ));
            } else {
                exists = true;
            }
        }
    }
    Ok(())
}

#[cfg(test)]
mod test {
    use super::*;

    use insta::assert_json_snapshot;

    #[test]
    fn snapshot_default_rules() {
        let rules = get_default_rules();

        assert_json_snapshot!("default_rules", rules);
    }

    #[test]
    fn snapshot_all_rules() {
        let rule_names = get_all_rule_names();

        assert_json_snapshot!("all_rule_names", rule_names);
    }

    #[test]
    fn verify_no_rule_properties_is_ok_when_empty() {
        let empty_properties = RuleProperties::default();

        assert_eq!(verify_no_rule_properties(&empty_properties), Ok(()));
    }

    #[test]
    fn verify_no_rule_properties_is_unexpected_rule_err() {
        let mut properties = RuleProperties::default();
        let some_rule_name = "rule name";
        properties.insert(some_rule_name.to_owned(), RulePropertyValue::None);

        assert_eq!(
            verify_no_rule_properties(&properties),
            Err(RuleConfigurationError::UnexpectedProperty(
                some_rule_name.to_owned()
            ))
        );
    }

    #[test]
    fn get_all_rule_names_are_deserializable() {
        for name in get_all_rule_names() {
            let rule: Result<Box<dyn Rule>, _> = name.parse();
            assert!(rule.is_ok(), "unable to deserialize `{}`", name);
        }
    }

    #[test]
    fn get_all_rule_names_are_serializable() {
        for name in get_all_rule_names() {
            let rule: Box<dyn Rule> = name
                .parse()
                .unwrap_or_else(|_| panic!("unable to deserialize `{}`", name));
            assert!(json5::to_string(&rule).is_ok());
        }
    }
}
