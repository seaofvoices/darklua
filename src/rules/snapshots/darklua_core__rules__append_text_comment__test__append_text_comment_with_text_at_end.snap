---
source: src/rules/append_text_comment.rs
expression: rule
---
{
  "rule": "append_text_comment",
  "location": "end",
  "text": "content"
}
